services/src/datasets/external/gbif.rs

```diff
@@ -0,0 +1,2467 @@
+pub struct GbifDataProviderDefinition {
+    name: String,
+    db_config: DatabaseConnectionConfig,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DataProviderDefinition for GbifDataProviderDefinition {
+    async fn initialize(self: Box<Self>) -> Result<Box<dyn DataProvider>> {
+        Ok(Box::new(GbifDataProvider::new(self.db_config).await?))
+    }
+
+    fn type_name(&self) -> &'static str {
+        "GBIF"
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DataProviderId {
+        GBIF_PROVIDER_ID
+    }
+}
+
+#[derive(Debug)]
+pub struct GbifDataProvider {
+    db_config: DatabaseConnectionConfig,
+    pool: Pool<PostgresConnectionManager<NoTls>>,
+}
+
+impl GbifDataProvider {
+    const LEVELS: [&'static str; 7] = [
+        "kingdom",
+        "phylum",
+        "class",
+        "order",
+        "family",
+        "genus",
+        "canonicalname",
+    ];
+
+    async fn new(db_config: DatabaseConnectionConfig) -> Result<Self> {
+        let pg_mgr = PostgresConnectionManager::new(db_config.pg_config(), NoTls);
+        let pool = Pool::builder().build(pg_mgr).await?;
+
+        Ok(Self { db_config, pool })
+    }
+
+    fn level_name(path: &str) -> String {
+        if path.is_empty() {
+            return Self::level_name_from_depth(0);
+        }
+        Self::level_name_from_depth(path.split('/').count())
+    }
+
+    fn level_name_from_depth(length: usize) -> String {
+        if length > Self::LEVELS.len() {
+            String::new()
+        } else {
+            Self::LEVELS[length].to_string()
+        }
+    }
+
+    fn get_filters(path: &str) -> Vec<(String, String)> {
+        if path.is_empty() {
+            vec![]
+        } else {
+            path.split('/')
+                .enumerate()
+                .filter_map(|(level, filter)| {
+                    if filter.eq("-") {
+                        None
+                    } else {
+                        Some((Self::level_name_from_depth(level), filter.to_string()))
+                    }
+                })
+                .collect()
+        }
+    }
+
+    fn extend_path(path: String, selection: &str) -> String {
+        if path.is_empty() {
+            selection.to_string()
+        } else {
+            path + "/" + selection
+        }
+    }
+
+    async fn get_datasets_items(
+        &self,
+        options: &Validated<LayerCollectionListOptions>,
+        path: &str,
+    ) -> Result<Vec<CollectionItem>> {
+        let taxonrank = path
+            .split_once('/')
+            .map_or_else(String::new, |(taxonrank, _)| taxonrank.to_string());
+        let path = path.split_once('/').map_or_else(|| "", |(_, path)| path);
+        let filters = GbifDataProvider::get_filters(path);
+        let conn = self.pool.get().await?;
+        let query = &format!(
+            r#"
+            SELECT DISTINCT canonicalname
+            FROM {schema}.species
+            WHERE taxonrank = $1{filter}
```

Comment:
They don't define it directly, but build it by navigating the hierarchy. The filter is part of the IDs. Since one could theoretically send anything as an ID, the filter string here uses prepared statement parameters to prevent exactly that: Building the ID such that SQL can be injected.
The column names, which cannot be parameterized, are checked against a whitelist.